### ISSUE TITLE
NADE Model Not Learning MNIST Properly

### ISSUE THREAD

**Issue by erdiphd** [EugenHotaj/pytorch-generative#41]
I tested the NADE model and trained it for 3500 epochs, but it did not learn the MNIST dataset properly. Below is my output.

![Image](https://github.com/user-attachments/assets/81430f85-caba-47b3-9616-dced66c94cdc)

I would like to understand the possible reasons for this. Could it be that 3500 epochs is still too low or is there a potential issue with the model itself?

Any insights would be greatly appreciated.

Thank you!

**Comment by EugenHotaj** [comment]
There may have been some changes with newer versions of pytorch that broke things. I have not run this code in a few years so hard to say off-hand what could be wrong, but 3500 epochs is probably way too many. Have you tried less than that, do the samples look wrong still?

My suggestion would be to look at the training/validation loss curves to see if they look reasonable. If they do, there may be some issues with the sampling code in newer version of pytorch, like some `dim=` args missing somewhere.

**Comment by erdiphd** [comment]
Thanks @EugenHotaj  for your answer.
I re-trained it again and the loss is discreasing shown as below;

![Image](https://github.com/user-attachments/assets/3bee694b-3df6-4439-bf0c-8ed84f1191bc)

However, the generated images doesnt look like the MNIST dataset. 

![Image](https://github.com/user-attachments/assets/60c636ef-17f2-4960-bc27-38e7f41fa21c)

I will try to investigate the issue and If I find something, I will update here.

**Comment by EugenHotaj** [comment]
@erdiphd yea if I had to guess there is probably something going wrong with the sampling code.
